crates/legacy/src/builder_state.rs

```diff
@@ -1143,3 +1141,414 @@ impl<TYPES: NodeType> BuilderState<TYPES> {
+        let transactions_2 = vec![TestTransaction::new(vec![2, 3, 4]); 2];
+        let (_quorum_proposal_2, quorum_proposal_msg_2, da_proposal_msg_2, builder_state_id_2) =
+            calc_proposal_msg(NUM_STORAGE_NODES, 0, None, transactions_2).await;
+
+        // process da proposal message first, so that later when process_da_proposal we can directly call `build_block` and skip storage
+        if let MessageType::DaProposalMessage(practice_da_msg_2) = da_proposal_msg_2.clone() {
+            builder_state
+                .process_da_proposal(practice_da_msg_2.clone())
+                .await;
+        } else {
+            panic!("Not a da_proposal_message in correct format");
+        }
+
+        // process quorum proposal, and do the check
+        if let MessageType::QuorumProposalMessage(practice_quorum_msg_2) =
+            quorum_proposal_msg_2.clone()
+        {
+            builder_state
+                .process_quorum_proposal(practice_quorum_msg_2.clone())
+                .await;
+        } else {
+            panic!("Not a quorum_proposal_message in correct format");
+        }
+
+        check_equal_quorum_proposal_hashmap(
+            builder_state.quorum_proposal_payload_commit_to_quorum_proposal,
+            correct_quorum_proposal_payload_commit_to_quorum_proposal,
+        );
+        // check global_state has this new builder_state_id
+        if let Some(_x) = global_state
+            .read_arc()
+            .await
+            .spawned_builder_states
+            .get(&builder_state_id_2)
+        {
+            tracing::debug!("global_state updated successfully");
+        } else {
+            panic!("global_state should have cooresponding builder_state_id as now we have matching da proposal.");
+        }
+    }
+
+    /// This test the function `process_decide_event`.
+    /// It checkes whether we exit out correct builder states when there's a decide event coming in.
+    /// This test also checks whether corresponding BuilderStateId is removed in global_state.
+    #[async_std::test]
+    async fn test_legacy_process_decide_event() {
+        async_compatibility_layer::logging::setup_logging();
+        async_compatibility_layer::logging::setup_backtrace();
+        tracing::info!("Testing the builder core with multiple messages from the channels");
+
+        // Number of views to simulate
+        const NUM_ROUNDS: usize = 5;
+        // Number of transactions to submit per round
+        const NUM_TXNS_PER_ROUND: usize = 4;
+        // Capacity of broadcast channels
+        const CHANNEL_CAPACITY: usize = NUM_ROUNDS * 5;
+        // Number of nodes on DA committee
+        const NUM_STORAGE_NODES: usize = 4;
+
+        // start builder_state without entering event loop
+        let (_senders, global_state, mut builder_state) =
+            start_builder_state_without_event_loop(CHANNEL_CAPACITY, NUM_STORAGE_NODES).await;
+
+        // insert some builder states
```

Comment:
This comment feels unrelated to any of the surrounded code
Is there anything it applies to or is it here by mistake?

Reply:
Changed to "// register some builder states for later decide event". This comment is about the following whole block, but I moved it to the place after transaction generation to avoid confusion. [1d14f1f](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/1d14f1f81eacfc86ec8baf62394d5ca28161c825)